Druid init action with automated test

adam-stasiak opened the pull request:
I added init action that starts Druid service. It can be easily verified with attached automated test. Ensure that you uploaded druid.sh file to your bucket and modified path inside test script (because current version expects druid.sh uploaded into official bucket).

functicons commented on druid/druid.sh:
```diff
@@ -0,0 +1,319 @@
+set -euxo pipefail
+
+export DRUID_VERSION='0.12.3'
+export DRUID_DIR="/root/druid-${DRUID_VERSION}"
```
Use $HOME, so that it could be run / tested by non-root.

functicons replied:
Use $HOME, so that it could be tested by non-root.

functicons commented on druid/druid.sh:
```diff
@@ -0,0 +1,319 @@
+}
+
+function download_druid() {
+  cd /root
```
ditto

functicons commented on druid/druid.sh:
```diff
@@ -0,0 +1,319 @@
+}
+
+function prepare_environment_services(){
+  systemctl start mysql
```
Document prerequisites and dependencies including mysql, zookeeper, etc in README.

functicons replied:
Is it possible to support Cloud SQL? https://github.com/GoogleCloudPlatform/dataproc-initialization-actions/tree/master/cloud-sql-proxy

functicons commented on druid/druid.sh:
```diff
@@ -0,0 +1,319 @@
+
+  bash /usr/lib/zookeeper/bin/zkServer.sh restart
+
+
```
nit: Remove extra empty lines.

functicons commented on druid/README.MD:
```diff
@@ -0,0 +1,33 @@
+This script installs Druid on dataproc clusters. To learn more about Druid see documentation: http://druid.io/docs/latest/design/.
+
+
+## Using this initialization action
```
Add a gcloud command line example.

functicons commented on druid/druid.sh:
```diff
@@ -0,0 +1,319 @@
+  bash bin/init
+
+  java \
+  -cp "lib/*" \
```
nit: Add indent of 4 spaces for continued lines.

functicons commented on druid/druid.sh:
```diff
@@ -0,0 +1,319 @@
+  systemctl enable coordinator overlord broker historical middleManager
+  systemctl start coordinator overlord broker middleManager historical
+
+}
```
Can we add "druid-" prefix for these service names? also use all lower case for service names, prefer "middle-manager" over "middleManager".

adam-stasiak commented:
@functicons Updated code - check this now. Now Druid init action can be used with cloud sql. I added automated tests for this but you must be patient because it takes a long time to execute them all :(

In Readme you can find steps how to create Druid cluster capable with Cloud SQL.

aniket486 commented on druid/druid.sh:
```diff
@@ -0,0 +1,320 @@
+  readonly enable_cloud_sql_metastore="$(/usr/share/google/get_metadata_value \
+    attributes/enable-cloud-sql-hive-metastore || echo 'true')"
+  local role=$(/usr/share/google/get_metadata_value attributes/dataproc-role)
+  if [[ ${role} == 'Master' ]]; then
```
We need to setup druid in following way:

Master:
Coordinator Nodes
Broker Nodes
Overlord Nodes

Workers:
Historical Nodes
Realtime Nodes
MiddleManager Nodes

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,320 @@
+
+  cat <<EOF > ${DRUID_DIR}/conf/druid/_common/common.runtime.properties
+#
+# Licensed to Metamarkets Group Inc. (Metamarkets) under one
```
I've just copied the original one. I will delete this.

adam-stasiak replied:
I removed Licence text.

adam-stasiak commented:
@aniket486 @functicons I applied suggestions.

aniket486 commented on druid/README.MD:
```diff
@@ -0,0 +1,54 @@
+## Using this initialization action
+You can use this initialization action to create a new Dataproc cluster with Druid installed.
+
+Once the cluster has been created, Druid Index Coordinator Console is configured to run on port `8090`. You can change this port with additional parameter:
```
Druid Console runs on port 8081 - may be that is more relevant here.

adam-stasiak replied:
Modified README with this:
`Once the cluster has been created, Druid Overlord Console is configured to run on port 8090. You can change this port with additional parameter: --metadata=druid-overlord-port=xxxx Druid Coordinator Console is configured to run on port 8081. You can change this port with additional parameter: --metadata=druid-coordinator-port=xxxx`

aniket486 commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+druid.indexing.doubleStorage=double
+EOF
+
+  cp /usr/lib/hadoop/etc/hadoop/core-site.xml conf/druid/_common/core-site.xml
```
better to symlink these rather than copying.
s/cp/ln -s

adam-stasiak replied:
Already fixed:
`  ln -sf /usr/lib/hadoop/etc/hadoop/core-site.xml conf/druid/_common/core-site.xml
  ln -sf /usr/lib/hadoop/etc/hadoop/hdfs-site.xml conf/druid/_common/hdfs-site.xml
  ln -sf /usr/lib/hadoop/etc/hadoop/yarn-site.xml conf/druid/_common/yarn-site.xml
  ln -sf /usr/lib/hadoop/etc/hadoop/mapred-site.xml conf/druid/_common/mapred-site.xml`

adam-stasiak commented:
@aniket486 I applied review suggestions. Now it should be compatible with GCS and JVM settings are calculated using equation from druid stacktrace (when it was not enough memory error).

Also removed cp and replaced them with symbolic links. All configuration stuff (how to change port, or use gcs) is described in README.

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,538 @@
+  else
+    num_merge_buff=$(max 2 ${num_threads}/4)
+  fi
+  xmx=$(calculate_xmx ${processing_buffer} ${num_threads} ${num_merge_buff})
```
This comes from stactrace error displayed when this value is not enough for druid. @aniket486

aniket486 commented on druid/druid.sh:
```diff
@@ -0,0 +1,538 @@
+
+  ln -sf /usr/share/java/mysql-connector-java.jar extensions/mysql-metadata-storage/mysql-connector-java-5.1.38.jar
+
+  java \
```
These two calls to `org.apache.druid.cli.Main tools pull-deps` can be folded together in one call.

java
 ...
 -c "org.apache.druid.extensions:mysql-metadata-storage:${DRUID_VERSION}" \
-c "org.apache.druid.extensions:druid-hdfs-storage:${DRUID_VERSION}"

We should also consider following: 
-c "org.apache.druid.extensions:druid-hdfs-storage:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions:mysql-metadata-storage:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions:druid-kafka-eight:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions.contrib:druid-google-extensions:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions.contrib:druid-distinctcount:${DRUID_VERSION}"

adam-stasiak replied:
@aniket486  I simplified this as you suggested.

adam-stasiak replied:
It's already made with this:
`  java \
    -cp "lib/*" \
    -Ddruid.extensions.directory="extensions" \
    -Ddruid.extensions.hadoopDependenciesDir="hadoop-dependencies" \
    org.apache.druid.cli.Main tools pull-deps \
    --no-default-hadoop \
    -c "org.apache.druid.extensions:druid-hdfs-storage:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions:mysql-metadata-storage:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions:druid-kafka-eight:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions.contrib:druid-google-extensions:${DRUID_VERSION}" \
    -c "org.apache.druid.extensions.contrib:druid-distinctcount:${DRUID_VERSION}"`

aniket486 commented:
@adam-stasiak - can you respond to the review comments above and resolve conversations wherever applicable?

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+
+  cat << 'EOF' > conf/druid/coordinator/jvm.config
+-server
+-Xms2g
```
Applied calculation

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+druid.indexer.logs.type=hdfs
+druid.indexer.logs.directory=/druid/indexing-logs
+
+# For S3:
```
Done

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+#druid.storage.bucket=your-bucket
+#druid.storage.baseKey=druid/segments
+#druid.s3.accessKey=...
+#druid.s3.secretKey=...
```
Done

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+
+# For HDFS (make sure to include the HDFS extension and that your Hadoop config files in the cp):
+druid.storage.type=hdfs
+druid.storage.storageDirectory=/druid/segments
```
now it is possible to make it running with GCS with metadata `gcs-bucket` with passed name of bucket space where all stuff will be uploaded. Init action continue configuration with additional actions for GCS.

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+
+  cd ${DRUID_DIR}
+
+  wget -O extensions/mysql-metadata-storage/mysql-connector-java-5.1.38.jar http://central.maven.org/maven2/mysql/mysql-connector-java/5.1.38/mysql-connector-java-5.1.38.jar
```
Fixed with symbolic link instead of download step.

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+
+
+  master_additional=$(/usr/share/google/get_metadata_value attributes/dataproc-master-additional)
+  cluster_name=$(hostname | sed -r 's/(.*)-[w|m](-[0-9]+)?$/\1/')
```
Applied.

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,511 @@
+  fi
+  echo | mysql -u root ${password_argument} -e "CREATE DATABASE druid DEFAULT CHARACTER SET utf8;"
+  echo | mysql -u root ${password_argument} -e "GRANT ALL ON druid.* TO 'druid' IDENTIFIED BY 'diurd';"
+  bash /usr/lib/zookeeper/bin/zkServer.sh restart
```
I added comment.

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,320 @@
+
+export DRUID_VERSION='0.12.3'
+export DRUID_HOME="/$(whoami)"
+export DRUID_DIR="${DRUID_HOME}/druid-${DRUID_VERSION}"
```
Druid home is changed.

adam-stasiak commented on druid/druid.sh:
```diff
@@ -0,0 +1,320 @@
+  readonly enable_cloud_sql_metastore="$(/usr/share/google/get_metadata_value \
+    attributes/enable-cloud-sql-hive-metastore || echo 'true')"
+  local role=$(/usr/share/google/get_metadata_value attributes/dataproc-role)
+  if [[ ${role} == 'Master' ]]; then
```
Now different services are installed on Master and Worker nodes.

adam-stasiak commented on druid/README.MD:
````diff
@@ -0,0 +1,53 @@
+Example of gcloud command for Druid initialization on standard cluster:
+
+```
+gcloud dataproc clusters create druid --master-machine-type n1-standard-4 --num-workers 2 --worker-machine-type n1-standard-4 --image-version 1.3-deb9 --initialization-actions 'gs://dataproc-initialization-actions/zookeeper/zookeeper.sh','gs://dataproc-initialization-actions/druid/druid.sh'
````
I added required info.

adam-stasiak commented:
@aniket486 I resolved all outstanding conversations and all suggestions are already in code.

aniket486 commented on druid/druid.sh:
```diff
@@ -0,0 +1,526 @@
+  else
+    num_merge_buff=$(max 2 ${num_threads}/4)
+  fi
+  xmx=$(calculate_xmx ${processing_buffer} ${num_threads} ${num_merge_buff})
```
Memory should be computed based on the size of nodes.

adam-stasiak replied:
@aniket486 Let me know what do you think about this simple change I've applied with last commit. I observed that when RAM size is lower than 7 GB for node then broker cannot be launched because lack of memory.

ChethanUK commented:
Any update on this?

medb commented:
> Any update on this?
Dataproc has [Druid component](https://cloud.google.com/dataproc/docs/concepts/components/druid), use it instead.

ChethanUK commented:
> > Any update on this?
> 
> Dataproc has [Druid component](https://cloud.google.com/dataproc/docs/concepts/components/druid), you can use it instead.

No flexibility, say I want to use the latest version of druid or some other use case..

medb commented:
> No flexibility, say I want to use the latest version of druid or some other use case..

We strive to install Druid and other software in init actions from our deb repository (the same that is used by optional components), so there no flexibility anyway.
